content/blog/prototyping/index.md

```diff
@@ -0,0 +1,755 @@
+
+The key is to keep things simple until it calls for more complexity.
+Start flat, then add structure incrementally as your understanding of the problem grows.
+
+See also [Matklad's article on large Rust workspaces](https://matklad.github.io/2021/08/22/large-rust-workspaces.html).
+
+## Summary
+
+Let's see how Rust stacks up against Python for prototyping:
+
+| Aspect                    | Python                                                                                                                             | Rust                                                                                                                      |
+| ------------------------- | ---------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------- |
+| Initial Development Speed | ✓ Very quick to write initial code<br>✓ No compilation step<br>✓ Simple syntax                                                     | ⚠️ Slightly slower initial development<br>✓ Type inference helps<br>✓ Tools like `bacon` provide quick feedback            |
+| Error Handling            | ❌ Silent failures possible<br>❌ Runtime exceptions<br>❌ Type errors discovered late                                                | ✓ Explicit error handling (with `unwrap` for prototypes)<br>✓ Compile-time error catching<br>✓ Type safety from the start |
+| Transition to Production  | ❌ Need extensive testing to catch type errors<br>❌ Often requires rewrite in another language<br>❌ Performance optimization needed | ✓ Same code can go to production<br>✓ Already has good performance<br>✓ Memory safety guaranteed                          |
+| Maintenance               | ❌ Type errors surface during runtime<br>❌ Refactoring is risky<br>❌ Documentation can get out of sync                              | ✓ Compiler catches most issues<br>✓ Safe refactoring with type system<br>✓ Types serve as documentation                   |
+| Debugging                 | ✓ Interactive debugger<br>✓ Easy print debugging                                                            | ✓ `dbg!` macro<br>✓ Compile-time catches most bugs                                                                                                 |
+| Code Evolution            | ❌ Hard to maintain large codebases<br>❌ Type issues compound                                                                      | ✓ Compiler guides improvements<br>✓ Types help manage complexity                         |
+| Collaboration             | ❌ Implicit assumptions about types<br>❌ Easy to break others' code<br>✓ Lower learning curve                                       | ✓ Explicit contracts through types<br>✓ Compiler prevents many conflicts<br>⚠️ Steeper learning curve                      |
```

Comment:
There are some, let's say "disingenuous" points against Python here.

- If type inference helps Rust, the dynamic typing in Python should be worth a point too.
- There absolutely are tools like `bacon` for Python, watching for files is not a Rust + Node oligopol.
- Rust has runtime panics too
- You wrote a whole chapter about `.unwrap()` in prototyping code, now you say "same code can go into production"?
- Documentation in Rust can go out of sync just the same as in Python.

Debugging, Code Evolution, and Collaboration seem very genuine. Maintenance except for the docs point too. But the others are a very bad comparison.

Reply:
Noted. Will update.

Reply:
Applied fixes